flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/DefaultExecutionDeployerTest.java

```diff
@@ -0,0 +1,388 @@
+
+        // deploy once to transition the tasks out from CREATED state
+        deployTasks(executionDeployer, executionGraph);
+
+        // The deploying of a non-CREATED vertex will result in IllegalStateException
+        try {
+            deployTasks(executionDeployer, executionGraph);
+            fail("IllegalStateException should happen");
+        } catch (IllegalStateException e) {
+            // expected exception
+        }
+    }
+
+    @Test
+    public void testDeploymentWaitForProducedPartitionRegistration() throws Exception {
+        shuffleMaster.setAutoCompleteRegistration(false);
+
+        final List<ResultPartitionID> trackedPartitions = new ArrayList<>();
+        partitionTracker.setStartTrackingPartitionsConsumer(
+                (resourceID, resultPartitionDeploymentDescriptor) ->
+                        trackedPartitions.add(
+                                resultPartitionDeploymentDescriptor
+                                        .getShuffleDescriptor()
+                                        .getResultPartitionID()));
+
+        final JobGraph jobGraph = nonParallelSourceSinkJobGraph();
+        final ExecutionGraph executionGraph = createExecutionGraph(jobGraph);
+        final ExecutionDeployer executionDeployer = createExecutionDeployer(executionGraph);
+
+        deployTasks(executionDeployer, executionGraph);
+
+        assertThat(trackedPartitions).hasSize(0);
+        assertThat(testExecutionOperations.getDeployedExecutions()).hasSize(0);
+
+        shuffleMaster.completeAllPendingRegistrations();
+        assertThat(trackedPartitions).hasSize(1);
+        assertThat(testExecutionOperations.getDeployedExecutions()).hasSize(2);
+    }
+
+    @Test
+    public void testFailedProducedPartitionRegistration() throws Exception {
+        shuffleMaster.setAutoCompleteRegistration(false);
+
+        final JobGraph jobGraph = nonParallelSourceSinkJobGraph();
+        final ExecutionGraph executionGraph = createExecutionGraph(jobGraph);
+        final ExecutionDeployer executionDeployer = createExecutionDeployer(executionGraph);
+
+        deployTasks(executionDeployer, executionGraph);
+
+        assertThat(testExecutionOperations.getFailedExecutions()).hasSize(0);
+
+        shuffleMaster.failAllPendingRegistrations();
+        assertThat(testExecutionOperations.getFailedExecutions()).hasSize(1);
+    }
+
+    @Test
+    public void testDirectExceptionOnProducedPartitionRegistration() throws Exception {
+        shuffleMaster.setThrowExceptionalOnRegistration(true);
+
+        final JobGraph jobGraph = nonParallelSourceSinkJobGraph();
+        final ExecutionGraph executionGraph = createExecutionGraph(jobGraph);
+        final ExecutionDeployer executionDeployer = createExecutionDeployer(executionGraph);
+
+        deployTasks(executionDeployer, executionGraph);
+
+        assertThat(testExecutionOperations.getFailedExecutions()).hasSize(1);
+    }
+
+    @Test
```

Comment:
Maybe add a Timeout for this test?

Reply:
Test timeout should be avoided. See "Avoid timeouts in JUnit tests" in https://flink.apache.org/contributing/code-style-and-quality-common.html#testing.

Reply:
OK